README.md
# React Reflex Layout
A simple implementation for flexbox components.

### Installation
npm i react-reflex-layout [--save]

### Example

```js
import React, { PropTypes, Component } from 'react';
import { FlexBox, FlexRow, FlexColumn, FlexItem, Spacer, asFlexItem } from 'react-reflex-layout';

import MyComponentHeader from '../MyComponentHeader';
import MyComponentFooter from '../MyComponentFooter';
import MySideBar from '../MySideBar';
import MyMainContent from '../MyMainContent';

const MySideBarAsFlexItem = asFlexItem()(MySideBar);

class MyComponent extends Component {

    render() {
        return (
            <FlexColumn flexBoxClass="MyComponent">
                <FlexItem flexItemClass="MyComponent-Header">
                    <MyComponentHeader />
                </FlexItem>
                
                <Spacer flexBasis="5px" flexGrow={0} flexItemClass="has-black-background"/>
                
                <FlexItem flexItemClass="MyComponent-Content">
                    <FlexBox flexDirection="row">
                        <MySideBarAsFlexItem flexBasis="20%" flexGrow={0} />
                                
                        <FlexItem>
                            <MyMainContent />
                        </FlexItem>
                    </FlexBox>
                </FlexItem>
                
                <FlexItem flexItemClass="MyComponent-Footer">
                    <MyComponentFooter />
                </FlexItem>
            </FlexColumn>
        );
    }
}
```

### Available Components
1. FlexBox - generic flexbox container
2. FlexRow - flexbox row container
3. FlexColumn - flexbox column container
4. FlexItem - flex item component
5. Spacer - an empty flex item

### Available decorators
1. asFlexItem - Wraps passed component in a FlexItem.


Refer code to see available properties for each component.


### Scripts
1. lint - npm run lint
2. test - npm test
3. build - npm run build


##### Warning
:warning: This module is pretty new and might have some bugs, please [file an issue](https://github.com/chetanism/react-reflex-layout/issues)
if you encounter any problems.

CHANGELOG.md
## React Reflex Layout Change Log




src/components/FlexItem.js
/**
 * Created by chetanv on 07/10/15.
 */

import React, { Component, PropTypes } from 'react';
import classNames from 'classNames';

class FlexItem extends Component {

    static propTypes = {

        flexItemClass: PropTypes.string,

        children: PropTypes.node.isRequired,

        order: PropTypes.number,
        flexGrow: PropTypes.number,
        flexShrink: PropTypes.number,
        flexBasis: function flexBasis(props, propName, componentName) {
            const val = props[propName];

            if (val === undefined || val === 'auto') {
                return null;
            }

            const intVal = parseInt(val, 10);

            if (isNaN(intVal) || intVal < 0) {
                return new Error(`Invalid flexBasis given: ${val} for component: ${componentName}`);
            }

            return null;
        },
    };

    static defaultProps = {
        flexItemClass: '',

        flexGrow: 1,
        flexShrink: 0,
        flexBasis: 'auto',
    };

    buildStyle() {
        const style = {
            position: 'relative',
        };

        if (this.props.order) {
            style.order = this.props.order;
        }

        if (this.props.flexGrow && this.props.flexGrow >= 0) {
            style.flexGrow = this.props.flexGrow;
        }

        if (this.props.flexShrink && this.props.flexShrink >= 0) {
            style.flexShrink = this.props.flexShrink;
        }

        if (this.props.flexBasis) {
            let val = this.props.flexBasis;

            if (val !== 'auto') {
                const intVal = parseInt(val, 10);

                if (val.indexOf('%') > -1) {
                    val = `${intVal}%`;
                } else {
                    val = `${intVal}px`;
                }
            }

            style.flexBasis = val;
        }

        return style;
    }

    render() {
        const style = this.buildStyle();
        return (
            <div className={ classNames('FlexItem', this.props.flexItemClass) } style={style}>
                {this.props.children}
            </div>
        );
    }
}

export default FlexItem;


src/components/FlexColumn.js
/**
 * Created by chetanv on 07/10/15.
 */

import React, { PropTypes, Component } from 'react';
import FlexBox from './FlexBox.js';

import classNames from 'classNames';

class FlexColumn extends Component {
    static propTypes = {
        flexBoxClass: PropTypes.string,

        reverse: PropTypes.bool,
        children: React.PropTypes.node.isRequired,
    };

    static defaultProps = {
        flexBoxClass: '',
        reverse: false,
    };

    render() {
        const { reverse, flexBoxClass, ...otherProps } = this.props;
        return (
            <FlexBox flexBoxClass={ classNames('FlexColumn', flexBoxClass) } flexDirection={ reverse ? 'column-reverse' : 'column' } {...otherProps} >
                {this.props.children}
            </FlexBox>
        );
    }
}

export default FlexColumn;


src/components/asFlexItem.js
/**
 * Created by chetanv on 07/10/15.
 */

import React, { PropTypes, Component } from 'react'; // eslint-disable-line no-unused-vars
import FlexItem from './FlexItem.js';

function asFlexItem() {
    return (ComposedComponent) => class extends Component {

        static propTypes = {
            children: PropTypes.node,
        };

        static defaultProps = {
            children: null,
        };

        render() {
            const { order, flexGrow, flexShrink, flexBasis, flexItemClass, ...others } = this.props;
            return (
                <FlexItem order={order} flexGrow={flexGrow} flexShrink={flexShrink} flexBasis={flexBasis} flexItemClass={flexItemClass}>
                    <ComposedComponent {...others}>
                        {this.props.children}
                    </ComposedComponent>
                </FlexItem>
            );
        }
    };
}

export default asFlexItem;


src/components/FlexBox.js
/**
 * Created by chetanv on 07/10/15.
 */

import React, { PropTypes, Component } from 'react';

import classNames from 'classNames';

class FlexBox extends Component {
    static propTypes = {
        children: PropTypes.node.isRequired,

        inline: PropTypes.bool,
        flexBoxClass: PropTypes.string,

        flexDirection: PropTypes.oneOf(['row', 'row-reverse', 'column', 'column-reverse']).isRequired,
        flexWrap: PropTypes.oneOf(['wrap', 'nowrap', 'wrap-reverse']),
        justifyContent: PropTypes.oneOf(['flex-start', 'flex-end', 'center', 'space-between', 'space-around']),
        alignItems: PropTypes.oneOf(['flex-start', 'flex-end', 'center', 'baseline', 'stretch']),
        alignContent: PropTypes.oneOf(['flex-start', 'flex-end', 'center', 'space-between', 'space-around', 'stretch']),
    };

    static defaultProps = {
        inline: false,
        flexBoxClass: '',

        flexWrap: 'nowrap',
        justifyContent: 'space-around',
        alignItems: 'stretch',
        alignContent: 'space-around',

    };

    buildStyle() {
        const style = {
            flexDirection: this.props.flexDirection,
            flexWrap: this.props.flexWrap,
            justifyContent: this.props.justifyContent,
            alignItems: this.props.alignItems,
            alignContent: this.props.alignContent,
        };

        const { inline } = this.props;

        if (inline) {
            style.display = 'inline-flex';
        } else {
            style.display = 'flex';
        }

        return style;
    }

    render() {
        const style = this.buildStyle();
        return (
            <div
                className={ classNames('FlexBox', this.props.flexBoxClass) }
                style={style}>
                {this.props.children}
            </div>
        );
    }
}

export default FlexBox;
